tests/test_contains.py

```diff
@@ -88,3 +121,125 @@ def test_cidr():
+  world = Group(key="world", name="world")
+  single_ip = Group(key="single_ip", name="single_ip")
+  includes_single_ip = Group(key="includes_single_ip", name="includes_single_ip")
+  group = Group(key="group", name="group")
+  world.save()
+  single_ip.save()
+  includes_single_ip.save()
+  group.save()
+
+  ip1 = IPRange(group=world, value="0.0.0.0/0")
+  ip1.save()
+  ip2 = IPRange(group=single_ip, value="192.168.0.2")
+  ip2.save()
+  
+  rel1 = Relation(subject=includes_single_ip, object=single_ip, relation=Relation.RelationType.INCLUSION)
+  rel1.save()
+  rel2 = Relation(subject=group, object=world, relation=Relation.RelationType.INCLUSION)
+  rel2.save()
+  rel3 = Relation(subject=group, object=includes_single_ip, relation=Relation.RelationType.EXCLUSION)
+  rel3.save()
+  
+  assert "192.168.0.2" not in group
+
+@pytest.mark.django_db
+def test_nested_inclusion():
+  """tests inclusion when there is a long chain of inclusions"""
+  group1 = Group(key="A", name="A")
+  group1.save()
+  group2 = Group(key="B", name="B")
+  group2.save()
+  group3 = Group(key="C", name="C")
+  group3.save()
+  group4 = Group(key="D", name="D")
+  group4.save()
+  ip = IPRange(group=group4, value="10.204.145.0")
+  ip.save()
+  relation2in1= Relation(
+    subject=group1, 
+    object=group2, 
+    relation=Relation.RelationType.INCLUSION
+    )
+  relation2in1.save()
+  relation3in1 = Relation(
+    subject=group1,
+    object=group3,
+    relation=Relation.RelationType.INCLUSION
+  )
+  relation3in1.save()
+  relation4in1= Relation(
+    subject=group1, 
+    object=group4, 
+    relation=Relation.RelationType.INCLUSION
+    )
+  relation4in1.save()
+  assert "10.204.145.0" in group1
+
+@pytest.mark.django_db
+def test_excluded_farther_down_chain():
+  """tests that the manager respects excluded IP addresses when 
+    excluded group is far down the chain, ticket 2280"""
+  ip = "129.2.19.99"
+  films_group = Group(key="films", name="Films@UM")
+  films_group.save()
+  shady_grove_group = Group(key="shady-grove-umfilms", name="Films@UM, Shady Grove")
+  shady_grove_group.save()
+  um_group = Group(key="um", name="UMD College Park campus and VPN")
+  um_group.save()
+  usmai_group = Group(key="usmai-ezproxy", name="USMAI EZProxy")
+  usmai_group.save()
+
+  ip_um = IPRange(group=um_group, value=ip)
+  ip_um.save()
+  ip_usmai = IPRange(group=usmai_group, value=ip)
+  ip_usmai.save()
+
+  relation_shady_in_films= Relation(
+    subject=films_group, 
+    object=shady_grove_group, 
+    relation=Relation.RelationType.INCLUSION
+    )
+  relation_shady_in_films.save()
+  relation_um_in_films= Relation(
+    subject=films_group, 
+    object=um_group, 
+    relation=Relation.RelationType.INCLUSION
+    )
+  relation_um_in_films.save()
+  relation_usmai_not_in_um= Relation(
+    subject=um_group, 
+    object=usmai_group, 
+    relation=Relation.RelationType.EXCLUSION
+    )
+  relation_usmai_not_in_um.save()
+
+  assert not ip in films_group
```

Comment:
This should be rewritten as `assert ip not in films_group`. (See also https://docs.python.org/3/reference/expressions.html#not-in)